21_midi2/main.go

```diff
@@ -0,0 +1,884 @@
+				prevX = x
+			}
+		}
+
+		// ジョイスティック Y 軸処理
+		{
+			y := ay.Get()
+			if 0x7000 <= y && y <= 0x9000 {
+				y = 0x8000
+			}
+			if y >= 0x8000 {
+				if prevY != y {
+					m.ControlChange(cable, channel, midi.CCModulationWheel, byte((y-0x8000)>>8))
+					prevY = y
+				}
+			}
+		}
+
+		// ロータリーエンコーダー位置変更時の処理
+		if newValue := rotaryEncoder.Position(); newValue != encOldValue {
+			// エンコーダーの変化方向を検出
+			if newValue > encOldValue {
+				// 右回転 - ドラムパターンを次へ
+				state.DrumPatternIndex = (state.DrumPatternIndex + 1) % len(drumPatterns)
+			} else {
+				// 左回転 - ドラムパターンを前へ
+				state.DrumPatternIndex = (state.DrumPatternIndex - 1 + len(drumPatterns)) % len(drumPatterns)
+			}
+
+			// ディスプレイ更新
+			redraw(display, state)
+
+			encOldValue = newValue
+		}
+
+		// ロータリーエンコーダーボタン処理
+		currentRotaryButton := rotaryButton.Get()
+		if prevRotaryButton && !currentRotaryButton {
+			// ボタンが押された
+			state.DrumPlaying = !state.DrumPlaying
+			// ディスプレイ更新
+			redraw(display, state)
+		}
+		prevRotaryButton = currentRotaryButton
+
+		// ドラムパターン再生処理
+		if state.DrumPlaying && state.DrumPatternIndex >= 0 && state.DrumPatternIndex < len(drumPatterns) {
+			currentPattern := drumPatterns[state.DrumPatternIndex]
+
+			// ノートオフ処理（前回の音を止める）
+			if !lastNoteOnTime.IsZero() && time.Since(lastNoteOnTime) >= 40*time.Millisecond && len(notesToTurnOff) > 0 {
+				for _, note := range notesToTurnOff {
+					m.NoteOff(cable, drumCh, midi.Note(note), 0)
+				}
+				notesToTurnOff = nil // クリア
+			}
+
+			// 次のステップを再生する時間になったか確認
+			if lastDrumTime.IsZero() || time.Since(lastDrumTime) >= time.Duration(currentPattern.StepLen)*time.Millisecond {
+				// 新しい音を鳴らす
+				notesToTurnOff = currentPattern.Steps[currentStep] // オフにする必要のあるノートを保存
+				for _, note := range notesToTurnOff {
+					m.NoteOn(cable, drumCh, midi.Note(note), velocity)
+				}
+				lastNoteOnTime = time.Now() // ノートオン時刻を記録
+
+				// 次のステップに進む
+				currentStep = (currentStep + 1) % len(currentPattern.Steps)
+				lastDrumTime = time.Now()
+			}
+		}
+
+		// キーの状態更新と処理
+		for i, s := range getKeys(colPins, rowPins) {
+			note := notes[i]
+			switch s {
+			case off2on:
+				m.NoteOn(cable, channel, note, velocity)
+
+				// 対応する色をLEDに設定
+				if color, exists := noteColors[note]; exists {
+					colors[i] = color
+				} else {
+					colors[i] = white // マッピングがない場合は白色を使用
+				}
+
+				// カタカナ音名を保存
+				if name, exists := noteNamesKatakana[note]; exists {
+					state.ActiveNotes[i] = name
+				} else {
+					state.ActiveNotes[i] = ""
+				}
+
+				state.Keys[i] = true
+				time.Sleep(1 * time.Millisecond)
+
+			case on2off:
+				m.NoteOff(cable, channel, note, velocity)
+
+				// LED の色をリセット
+				colors[i] = black
+
+				// 音名をクリア
+				state.ActiveNotes[i] = ""
+				state.Keys[i] = false
+				time.Sleep(1 * time.Millisecond)
+			}
+		}
+
+		// LED に色を反映
+		ws.WriteRaw(colors)
+
+		// 画面を更新
+		redraw(display, state)
+
+		if cnt > 0 && cnt%10 == 0 {
+			time.Sleep(1 * time.Millisecond)
+		}
+		cnt++
```

Comment:
counter を使う案を出しましたが、ここも cnt よりも time.Since() を使う形にして 10ms を刻むとよさそう。
なお、現時点のコードは毎周期 redraw してしまうので、あまりよくないと思います。
※display に対して処理が速すぎるので、 LCD がフリーズしちゃいそう

Reply:
修正できてるので閉じ。